include/ncengine/graphics/SkeletalAnimator.h

```diff
@@ -13,38 +13,96 @@ namespace nc::graphics
+        auto AddState(const anim::Loop& loopProperties) -> uint32_t;
+
+        /** 
+         * @brief Add an anim::PlayOnce state to the state machine.
+         * @param playOnceProperties Properties required to create an anim::PlayOnce state.
+         */
+        auto AddState(const anim::PlayOnce& playOnceProperties) -> uint32_t;
+
+        /** 
+         * @brief Add an anim::Stop state to the state machine.
+         * @param stopProperties Properties required to create an anim::Stop state.
+         */
+        auto AddState(const anim::Stop& stopProperties) -> uint32_t;
+
+        /**
+         * @brief Interrupt the state machine with an immediate looping animation.
+         * Transitions back into the state machine when exitWhen is met.
+         * @param animUid The animation to play.
+         * @param exitWhen Exit back to the state machine when this condition is met.
+         * @param exitTo Exit back to this state.
+         */
+        void LoopImmediate(const std::string& animUid, const std::function<bool ()>& exitWhen, uint32_t exitTo);
+
+        /**
+         * @brief Interrupt the state machine with an immediate animation that plays once.
+         * Transitions back into the state machine when the animation has completed.
+         * @param animUid The animation to play
+         * @param exitTo Exit back to this state.
+         */
+        void PlayOnceImmediate(const std::string& animUid, uint32_t exitTo);
+
+        /**
+         * @brief Interrupt the state machine and stop the currently playing animation.
+         * Transitions back into the state machine when exitWhen is met.
+         * @param exitWhen Exit back to the state machine when this condition is met.
+         * @param exitTo Exit back to this state.
+         */
+        void StopImmediate(const std::function<bool ()>& exitWhen, uint32_t exitTo);
+
+        /** @brief Get the first state in the state machine, initialized by the constructor. */
+        auto RootState() -> uint32_t { return m_states.Get(0)->id; }
+
         /** @brief Connect to SkeletalAnimationSystem. **/
-        auto Connect() noexcept -> Signal<const anim::PlayState&>& { return m_onPlayStateChanged; }
+        auto Connect() noexcept -> Signal<const anim::StateChange&>& { return m_onStateChanged; }
 
-        /** @brief Get UID of mesh that will be animated **/
+        /** @brief Get the UID of the mesh that will be animated **/
         auto MeshUid() const noexcept -> const std::string& { return m_meshUid; }
 
         /** @brief Used to manage the play once state **/
         void CompleteFirstRun();
 
     private:
-        auto ExecuteInitialState() -> bool;
+        auto ExecuteRootState() -> bool;
+        auto ExecuteImmediateState() -> bool;
+        auto ExecuteExitState() -> bool;
+        auto ExecuteChildEnterState(uint32_t childStateHandle) -> bool;
+        auto AddState(anim::StopImmediate stopImmediateProperties) -> uint32_t;
 
+        anim::SparseSet m_states;
```

Comment:
This is 408 bytes or 6 cache lines. Member order will help a bit, but the killer is the `std::optional<anim::State>`. IIRC the immediate state is sort of a one-off, rare thing. If so, we could put it behind a `std::unique_ptr`. 168 bytes is feasible with minimal changes. We could also consider a cold pointer, but we'd probably want to land below 128 bytes to make it worth it; requires some benchmarking. That might be better left for a separate issue.